zson/formatter.go

```diff
@@ -285,42 +285,61 @@ func (f *Formatter) formatVector(indent int, open, close string, inner zed.Type,
 	indent += f.tab
 	sep := f.newline
 	it := zv.Iter()
-	union, _ := zed.TypeUnder(inner).(*zed.TypeUnion)
-	seen := make(map[zed.Type]struct{})
+	elems := newElemBuilder(inner)
 	for !it.Done() {
 		f.build(sep)
 		f.indent(indent, "")
-		b := it.Next()
-		typ := inner
-		if union != nil {
-			if b == nil {
-				// The type returned from union.SplitZNG for a null value will
-				// be the union type. While this is the correct type, for
-				// display purposes we do not want to see the decorator so just
-				// set the type to null.
-				typ = zed.TypeNull
-			} else {
-				typ, b = union.SplitZNG(b)
-				if _, ok := seen[typ]; !ok {
-					seen[typ] = struct{}{}
-				}
-			}
-		}
+		typ, b := elems.add(it.Next())
 		if err := f.formatValue(indent, typ, b, known, parentImplied, true); err != nil {
 			return true, err
 		}
 		sep = "," + f.newline
 	}
 	f.build(f.newline)
 	f.indent(indent-f.tab, close)
-	if _, isnamed := inner.(*zed.TypeNamed); union != nil && (isnamed || len(seen) != len(union.Types)) {
+	if elems.needsDecoration() {
 		// If we haven't seen all the types in the union, print the decorator
 		// so the fullness of the union is persevered.
 		f.decorate(zv.Type, false, true)
 	}
 	return false, nil
 }
 
+type elemBuilder struct {
```

Comment:
I don't know what to call this but it isn't a builder.

Reply:
This is true.